### PR TITLE
Add selective import symbols to dtemplate.d

### PR DESCRIPTION
**Pull request by SamAxe**
I was looking at the dtemplate.d file to understand it a better in the context of ongoing mailing list conversations and had done this locally and folks on the beerconf meeting suggested that I make a pull request.

I ran 
./build.d style
./build.d
./build.d unittest

I ran make -f posix.mak unittest in phobos and everything passed.

Tested on Macbook pro with Ubuntu 20.04

If this type of patch is desired, I can probably do a few more.

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Nice. A few comments:

- Please use a more expressive commit message (e.g. smth like `Use selective imports in dtemplate.d`) + maybe a message
- For the future: it would be interesting to make some of those imports local if possible
- Please ignore the Azure failure, that's a temporary, unrelated problem

**Comment by SamAxe** [comment]
If you'd prefer to reject this PR, I can make a new pull request with better commit message and explore the local imports.  I thought the 1:1 so far might be incrementally small delta change.

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -38,43 +38,54 @@
...
+import dmd.declaration : MatchAccumulator, Declaration, VarDeclaration, STC, AliasDeclaration,
+    OverDeclaration, TupleDeclaration;
+import dmd.dmangle : mangleToBuffer;
+import dmd.dmodule : Module;
+import dmd.dscope : Scope, SCOPE;
+import dmd.dsymbol : Dsymbol, ScopeDsymbol, WithScopeSymbol, Prot, DsymbolTable, PASS, Ungag,
+    OverloadSet, foreachDsymbol;
+import dmd.dsymbolsem : aliasSemantic, dsymbolSemantic, templateInstanceSemantic;
+import dmd.errors : error, errorSupplemental, fatal, message, tip;
+import dmd.expression : Expression, VarExp, IntervalExp, RealExp, ComplexExp, IdentityExp, NullExp,
```
Comment (MoonlightSentinel) [inline src/dmd/dtemplate.d:58]: Maybe these import should not be selective given that the import iterates most of the symbols defined in `expression.d`.

Comment (SamAxe) [reply src/dmd/dtemplate.d:58]: I had that thought too.  I'm just trying to feel out what is preferred, useful, etc.

**Comment by MoonlightSentinel** [comment]
> If you'd prefer to reject this PR, I can make a new pull request with better commit message and explore the local imports. I thought the 1:1 so far might be incrementally small delta change.

This step is fine by me, you can change the commit message using

```shell
git commit --amend
git push --force
```

**Comment by SamAxe** [comment]
Sounds good.  There is more beerconf left. :)

**Review by Geod24 — CHANGES_REQUESTED** [review]
I don't think we should merge this. The selective imports either import most of the symbols in the modules (see mtype), or import the "main" symbol in the module (e.gg. `dmodule`, `dscope`). Hence, I don't see much advantage in the selective import.

Additionally, this will confuse new contributors: Imagine ONE file having only selective imports, and all others being regular imports. Finally, if we are to follow that style only in this module (which, for obvious reason, doesn't make sense), it would be an annoyance every time something is added.

I hope this doesn't discourage you from contributing though! You should also join our Slack, if you're not already there.

**Comment by MoonlightSentinel** [comment]
> I don't think we should merge this. The selective imports either import most of the symbols in the modules (see mtype), or import the "main" symbol in the module (e.gg. `dmodule`, `dscope`). Hence, I don't see much advantage in the selective import.
 
Agreed on `mtype` and `expression`.

> Additionally, this will confuse new contributors: Imagine ONE file having only selective imports, and all others being regular imports. Finally, if we are to follow that style only in this module (which, for obvious reason, doesn't make sense), it would be an annoyance every time something is added.

Well, you'll have to start somewhere to make a change. Regardless, I think the real value lies in using local imports whenever possible to endorse stronger encapsulation..

**Comment by SamAxe** [comment]
The responsiveness and review on this PR very nice.  Thank you.  A little more follow up to see what direction this should take.

1. Would it be more desirable to convert the entire dmd directory with a change set?  

2. For really long functions...If symbol used in single place in block, should it go there rather than at function level?

3. How many times should does a symbol used in a file suggest global vs local imports?  
    Maybe 2 or 3?  Or just eliminate global in general except for some very frequently used symbols, e.g. print?

4. If symbol is used in single function argument variable, should the import move to the function or leave global?
4a.  Seems there was some template pattern to address this, should it be used?

- I'm on the IRC channel.  What do I need to do to join Slack?

**Pull request closed without merging**